_posts/2023-10-11-kafka-consumer-client-essentials.md

```diff
@@ -0,0 +1,485 @@
+        kafkaConsumer.commitSync(pendingOffsets);
+        pendingOffsets.clear();
+    }
+    
+    @Override
+    public void onPartitionsLost(Collection<TopicPartition> partitions) {
+        System.out.printf("Lost partitions: {}", partitions);
+    }
+
+    @Override
+    public void onComplete(Map<TopicPartition, OffsetAndMetadata> map, Exception e) {
+        if (e != null) {
+            System.err.println("Failed to commit offsets");
+            if (!retriable(e)) {
+                e.printStackTrace();
+                System.exit(1);
+            }
+        }
+    }
+}
+```
+
+### Running the consumer application
+
+To put this client into action, simply run the main method in the `consumer` class. 
+The client consumes messages from the specified topic (`TOPIC_NAME`) until it reaches the predefined message count, which is 50 messages with the `NUM_MESSAGES` constant value we specified. 
+
+> The consumer is not designed to be safely accessed concurrently by multiple threads. Ensure that only one instance of the consumer application is running to maintain the expected behavior.
+
+### Error handling
+
+When developing a Kafka consumer application, it's important to consider how you want it to handle different types of exceptions. 
+
+The error handling capabilities we introduced ensures that the consumer application can recover from certain retriable errors while addressing others as non-retriable, terminating operation of the client when necessary. 
+
+Here's a breakdown of retriable and non-retriable errors that the client handles:
+
+**Non-retriable errors caught by the consumer application**
+
+* `InterruptedException`: This error occurs when the current thread is interrupted while paused. 
+Interruption typically happens during consumer shutdown or when stopping its operation. 
+The exception is rethrown as a `RuntimeException`, which ultimately terminates the consumer.
+
+* `IllegalArgumentException`: This error is thrown when the consumer receives invalid or inappropriate arguments. 
+For instance, it can be triggered if essential details like the topic are missing.
+
+* `UnsupportedOperationException`: This error is raised when an operation is not supported or when a method is not implemented as expected. 
+For instance, it can be triggered if you attempt to use an unsupported consumer configuration, leading to a runtime exception.
+
+**Retriable errors caught by the consumer application**
+
+`RetriableException`: This type of error is thrown for any exception that implements the `RetriableException` interface, as provided by the Kafka client library.
+
+`OffsetOutOfRangeException`: This error is thrown when the consumer attempts to seek to an invalid offset for a partition, typically when the offset is outside the valid range of offsets for that partition. To recover, the consumer seeks to the end of the partition to commit the offset synchronously (`commitSync`).
```

Comment:
Isn't the behavior controlled by auto.offset.reset? This is set to `earliest` so it should seek to the _start_ of the partition.
Same below

Reply:
@fvaleri - does auto.offset.reset control the behavior here? and for `NoOffsetForPartitionException`. That is should we be saying _start_ here?

Reply:
Yes. The original comment I had inside that catch clause was "// invalid or no offset found without auto.reset.policy", which I think we should keep and also put in the exception description. Users are expected to play with this example, and if they set auto.offset.reset: none, then they are covered.

Reply:
Thanks. I've described what happens with these errors when auto-reset policy is and isn't enabled